api/api/versions_api.go

```diff
@@ -145,6 +146,108 @@ func (c *VersionsController) CreateVersion(r *http.Request, vars map[string]stri
 	return Created(version)
 }
 
+func (c *VersionsController) DeleteVersion(r *http.Request, vars map[string]string, _ interface{}) *Response {
+	ctx := r.Context()
+
+	modelID, _ := models.ParseID(vars["model_id"])
+	versionID, _ := models.ParseID(vars["version_id"])
+
+	model, version, err := c.getModelAndVersion(ctx, modelID, versionID)
+	if err != nil {
+		if !gorm.IsRecordNotFoundError(err) {
+			return InternalServerError(err.Error())
+		}
+		return NotFound(err.Error())
+	}
+
+	// handle for pyfunc v2, since the prediction job feature is only available for pyfunc_v2 model
+	if model.Type == "pyfunc_v2" {
```

Comment:
Since much of this logic (and `checkActivePredictionJobs` and `checkActiveEndpoints`) is largely this same as with the model delete, can we extract it into a private method of the `VersionsController` and let model delete call it too? In this case, the `ModelsController` will hold a reference to the `VersionsController` like:

```golang
// ModelsController controls models API.
type ModelsController struct {
	*AppContext
        *VersionsController
}

func (c *ModelsController) DeleteModel(r *http.Request, vars map[string]string, body interface{}) *Response {
    for _, version := range versions {
            // This function can be defined in versions_api.go and handle
            // deletion of a single version / mlflow run
            err := c.VersionsController.deleteVersion(version)
            if err != nil {
                    // Handle type of error and return Response 
            }
    }

    // Delete mlflow experiment and artifact
    // ...

    // Delete model data from DB
    // ...

    return Ok(model.ID)
}
```

Wdyt?

Reply:
I think it's safer to perform first a check on all the versions (in a loop) to determine if there are any versions that have active endpoint/prediction jobs. If none of them are active, we can then proceed to delete them in a loop.

Reply:
Sorry I missed this discussion earlier. Let's keep this implementation for now.